File: src/console/screen/impl/SetupScreen.php
<?php

namespace pocketcloud\cloud\console\screen\impl;

use pocketcloud\cloud\console\Console;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\console\log\logger\ILogger;
use pocketcloud\cloud\console\log\output\SetupOutputHandler;
use pocketcloud\cloud\console\screen\IScreen;
use pocketcloud\cloud\setup\Setup;

final class SetupScreen extends IScreen {

    private ?SetupOutputHandler $outputHandler = null;
    private ?ILogger $logger = null;

    public function __construct(private readonly Setup $setup) {}

    public function initialize(Console $console): void {
        $this->clearConsole();
        $this->disableHistory();
        $this->setControlCHandler(fn() => $this->setup->cancel());
        $this->setCompletionHandler(function (array $tokens, string $current): array {
            $recommendations = $this->setup->getCurrentQuestion()?->getPossibleAnswers() ?? [];
            if (empty($recommendations) || !empty($tokens)) return [];
            $matches = [];
            foreach ($recommendations as $recommendation) {
                if (str_starts_with(strtolower($recommendation), strtolower($current))) {
                    $matches[] = $recommendation;
                }
            }

            return $matches;
        });

        $this->logger = CloudLogger::tmp();
        $this->logger->setFormat("§r{message}");

        $this->setOutputHandler($this->outputHandler = new SetupOutputHandler());
        $this->outputHandler->addAuthorizedLogger($this->logger);
    }

    public function tick(int $currentTick): void {}

    public function handleInput(string $input): void {
        $this->setup->handleInput($input);
    }

    public function onRemove(int $currentTick): void {
        $this->clearConsole();
        $this->enableHistory();
        $this->resetOutputManager();
        $this->setInput("");
        $this->restoreAll();
        $this->printLogCache();
    }

    public function getOutputHandler(): ?SetupOutputHandler {
        return $this->outputHandler;
    }

    public function getLogger(): ?ILogger {
        return $this->logger;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/screen/impl/SetupScreen.java
package de.pocketcloud.cloud.console.screen.impl;

import de.pocketcloud.cloud.console.CloudConsole;
import de.pocketcloud.cloud.console.screen.Screen;
import de.pocketcloud.cloud.console.util.InterruptionResult;
import de.pocketcloud.cloud.setup.Question;
import de.pocketcloud.cloud.setup.Setup;
import org.jline.reader.Candidate;
import org.jline.reader.LineReader;
import org.jline.reader.ParsedLine;

import java.util.ArrayList;
import java.util.List;

public final class SetupScreen extends Screen {

    private final Setup setup;

    public SetupScreen(Setup setup) {
        this.setup = setup;
    }

    @Override
    public void initialize(CloudConsole console) {
        clear();
        console.disableHistory();
        enableCompletion();
    }

    @Override
    public void handleInput(String input) {
        setup.handleInput(input);
    }

    @Override
    public void tick(long currentTick) {}

    @Override
    public void onRemove(long currentTick) {
        clear();
        restoreAll();
        printLogCache();
    }

    @Override
    public InterruptionResult onCancel(long currentTick) {
        setup.cancel();
        return InterruptionResult.CONTINUE;
    }

    @Override
    public void onTabComplete(LineReader reader, ParsedLine parsedLine, List<Candidate> list) {
        Question<?> question = setup.getCurrentQuestion();
        if (question != null) {
            final List<String> words = new ArrayList<>(parsedLine.words());
            if (!question.getPossibleAnswers().isEmpty()) {
                if (words.isEmpty() || words.size() == 1) {
                    String current = words.isEmpty() ? "" : words.getFirst();
                    for (String possibleAnswer : question.getPossibleAnswers()) {
                        if (possibleAnswer.toLowerCase().startsWith(current)) {
                            list.add(new Candidate(possibleAnswer));
                        }
                    }
                }
            }
        }
    }
}